### PR TITLE
Two Step Authentication (or One Token to Rule them All)

### LINKED ISSUE #2298: Improve WP.com Two-Factor Authentication
Make two-factor authentication easier:

 * Remove the requirement for an app-specific password, instead allow entry of the time-based code from  an auth app or sms.
 * Maybe integrate with the auth apps to see if we can get the code instantly (not sure if authy/google have available intents for something like that)
 * Implement #911 so that xml-rpc requests will still work.

### PR DESCRIPTION
**Pull request by roundhill**
Adds the ability to sign in to the app using a time-based code (instead of creating an application specific password) when signing in to the app. This makes things way more convenient, but it also opens up the opportunity to not store the password in the app any longer for WordPress.com accounts. That opportunity if why this PR became quite large...

When the app updates to the new `WordPressDB` version, it will clear out the password for all WP.com accounts, since the app will now use the token for authorization for all network calls.

The token is used to authorize XML-RPC calls now (#911) and is also sent to authorize a user when we load a `WPWebViewActivity`. The token is also now viewed as the ultimate sign of being authorized to WP.com (instead of previously when we'd check for a username and password). If the app receives an invalid token response from the REST API, a `SignInActivity` is immediately shown.

I ended up touching areas of the app typically maintained by other devs. Here's some details:

* In the Reader, the `ReaderWebView` now intercepts image requests and re-requests the image with the auth token in the header. This also removed the need for `ReaderAuthActions` class, so I removed it. cc @nbradbury 

* In Stats, we now show the `SignInActivity` to authorize Jetpack sites, in order to support two-step authentication there as well. cc @daniloercoli 

* Since this touches a lot of core areas as well as the Sign In process... cc @maxme :) I had to change how the new user process worked, but I've tested it a bunch and it still works.

I made a video that helps explain the changes visually: https://cloudup.com/ijM50wKovOd

Includes changes to the `android-wordpress-com-rest` and `WordPress-Networking-Android` libraries. Fixes #2298 and #911.

### CODE REVIEW

**Comment by nbradbury** [comment]
> This also removed the need for ReaderAuthActions class

So happy about this.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderWebView.java [hunk 1]
```diff
@@ -188,6 +207,34 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
                 return false;
             }
         }
+
+        @Override
+        public WebResourceResponse shouldInterceptRequest(WebView view, String url) {
+            // Intercept requests and add the WP.com authorization header
+            if (!TextUtils.isEmpty(mToken) && isImageUrl(url)) {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderWebView.java]: It seems inefficient to do this for every image when most posts aren't private. How about we add a `hasPrivateImages` property to ReaderWebView that determines whether this is necessary? ReaderPostDetailFragment would set that property [here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/reader/ReaderPostDetailFragment.java#L721) based on `mPost.isPrivate`.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderWebView.java [hunk 2]
```diff
@@ -188,6 +207,34 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
...
+            }
+
+            return super.shouldInterceptRequest(view, url);
+        }
+    }
+
+    private static boolean isImageUrl(String url) {
+        if (TextUtils.isEmpty(url)) return false;
+
+        String lowerCaseUrl = url.toLowerCase();
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderWebView.java]: This comparison will return false for image URLs that have a query string. You can use `UrlUtils.removeQuery` to remove the query string.

File: WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderWebView.java [hunk 3]
```diff
@@ -188,6 +207,34 @@ public boolean shouldOverrideUrlLoading(WebView view, String url) {
...
+
+        return lowerCaseUrl.endsWith("jpg") || url.endsWith("jpeg") || url.endsWith("gif") ||
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderWebView.java]: Only the first comparison is using `lowerCaseUrl` - the rest incorrectly use `url`.

**Comment by nbradbury** [comment]
Assorted feedback after using 2fa signin:
* The link to send the code via SMS doesn't have a pressed state
* The app calls it a "verification code" but the text message you get calls it a "security code"
* You can still edit the verification code while sign in is ongoing
* "Enter the code from your authenticator app or get the code via text message" may not make sense to those who don't know what 2fa is all about. What do you think of using a separate screen for the verification code which explains it a bit more?
* There should be a comment next to [WPCOM_PASSWORD_PREFERENCE](https://github.com/wordpress-mobile/WordPress-Android/blob/issue/2298-2fa-improvements/WordPress/src/main/java/org/wordpress/android/WordPress.java#L80) explaining that it's no longer used
* This looks harmless, but we still write the (empty) password pref in [migrateWPComAccount](https://github.com/wordpress-mobile/WordPress-Android/blob/issue/2298-2fa-improvements/WordPress/src/main/java/org/wordpress/android/WordPressDB.java#L284)

File: WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderWebView.java (hunk 1 above)
Comment (roundhill) [inline WordPress/src/main/java/org/wordpress/android/ui/reader/views/ReaderWebView.java]: Sounds good, I didn't know we had the ability to tell if the post is private :+1:

**Comment by roundhill** [comment]
> The link to send the code via SMS doesn't have a pressed state
Yeah... I took it off because I couldn't get just the underlined text to highlight. Think it's ok to highlight the entire button?

> "Enter the code from your authenticator app or get the code via text message" may not make sense to those who don't know what 2fa is all about.
The user has to explicitly enable 2fa, so they should have some idea about it... but I see your point. I like having it on the same login screen vs. a separate screen. They can always click the help button and get support.

**Comment by nbradbury** [comment]
> Think it's ok to highlight the entire button?

That would probably look weirder than no highlight. Does the link highlight on press if you make it an anchor tag rather than just [underline](https://github.com/wordpress-mobile/WordPress-Android/blob/issue/2298-2fa-improvements/WordPress/src/main/java/org/wordpress/android/ui/accounts/SignInFragment.java#L187) it?

**Comment by roundhill** [comment]
> The app calls it a "verification code" but the text message you get calls it a "security code"

This has been fixed server-side.

**Comment by roundhill** [comment]
> Does the link highlight on press if you make it an anchor tag rather than just underline it?

No, didn't work. I could split the text into two sentences, then I could make it work as we'd expect it to?

Something like: `Enter the code from your authenticator app. Send code via text message.`

**Comment by nbradbury** [comment]
> I could split the text into two sentences

That sounds like a good approach.

File: WordPress/src/main/java/org/wordpress/android/ui/accounts/BlogUtils.java [hunk 4]
```diff
@@ -26,12 +26,11 @@
...
      * @return true if a change has been made (new blog added, old blog updated, blog deleted).
      */
-    public static boolean syncBlogs(Context context, List<Map<String, Object>> newBlogList, String username,
-                                    String password, String httpUsername, String httpPassword) {
+    public static boolean syncBlogs(Context context, List<Map<String, Object>> newBlogList, String username) {
         boolean retValue;
 
         // Add all blogs from blogList
-        retValue = addBlogs(newBlogList, username, password, httpUsername, httpPassword);
+        retValue = addBlogs(newBlogList, username, null, null, null);
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/accounts/BlogUtils.java]: reuse `addBlogs(List<Map<String, Object>> userBlogList, String username)` defined below?

**Comment by maxme** [comment]
* We should make the username/password viewrs (+code textview) inactive when it's verifying the code, or we can result in: https://cloudup.com/cVAAJqQdkLd
* Can we hide the "Send code via text message" button if the user has not filled his phone number?
* Bug: make the code textview visible, then tap help, go back and try to login again: https://cloudup.com/cJyxEkNbURG

**Comment by roundhill** [comment]
@maxme Those UI bugs should be fixed, good catch!

> Can we hide the "Send code via text message" button if the user has not filled his phone number?

Every user that has enabled two-step should have entered their phone number during the setup process at WP.com.

**Comment by maxme** [comment]
> Every user that has enabled two-step should have entered their phone number during the setup process at WP.com.

Ah, I don't remember having entered a phone number on my test account! Someone has received a lot of codes this morning ;)

**Comment by daniloercoli** [comment]
I did lot of testing (login/logout, different devices, invalidating tokens, etc) and everything seems ok.

**Comment by maxme** [comment]
Looks good :shipit:

**Comment by maxme** [comment]
shipped!

**Pull request merged**